File: core/src/main/java/pcb/uwu/amount/composite/mechanics/Acceleration.java
package pcb.uwu.amount.composite.mechanics;

import pcb.uwu.amount.quantity.Mass;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.composite.mechanics.AccelerationUnit;
import pcb.uwu.units.composite.mechanics.ForceUnit;

import java.math.BigDecimal;
import java.math.MathContext;

public class Acceleration extends CompositeUnitAmount<AccelerationUnit> {

	// region constructors

	public Acceleration(Number value, AccelerationUnit unit) {
		super(value, unit);
	}

	public Acceleration(Number value, Magnitude magnitude, AccelerationUnit unit) {
		super(value, magnitude, unit);
	}

	public Acceleration(String value, AccelerationUnit unit) {
		super(value, unit);
	}

	public Acceleration(String value, Magnitude magnitude, AccelerationUnit unit) {
		super(value, magnitude, unit);
	}

	public Acceleration(BigDecimal value, AccelerationUnit unit) {
		super(value, unit);
	}

	public Acceleration(BigDecimal value, Magnitude magnitude, AccelerationUnit unit) {
		super(value, magnitude, unit);
	}

	public Acceleration(BigDecimalAmount amount, AccelerationUnit unit) {
		super(amount, unit);
	}

	public Acceleration(BigDecimalAmount amount, Magnitude magnitude, AccelerationUnit unit) {
		super(amount, magnitude, unit);
	}
	
	// endregion

	// region implement UnitAmount

	@Override
	public Acceleration plus(UnitAmount<AccelerationUnit> other, MathContext mathContext) {
		return new Acceleration(plusAmount(other, mathContext), getUnit());
	}

	@Override
	public Acceleration minus(UnitAmount<AccelerationUnit> other, MathContext mathContext) {
		return new Acceleration(minusAmount(other, mathContext), getUnit());
	}

	@Override
	public Acceleration multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Acceleration(multipliedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Acceleration dividedBy(BigDecimal other, MathContext mathContext) {
		return new Acceleration(dividedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Acceleration convertTo(AccelerationUnit unit) {
		return new Acceleration(getAmountIn(unit), unit);
	}

	// endregion

	// region composition

	public Force multipliedBy(Mass mass, MathContext mathContext) {
		BigDecimalAmount amount = getAmount()
				.multipliedBy(mass.getAmount().getValue(), mathContext);
		ForceUnit unit = new ForceUnit(mass.getUnit(), getUnit());

		return new Force(amount, unit);
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/amount/derived/mechanics/Acceleration.kt
package pcb.uwu.amount.derived.mechanics

import pcb.uwu.amount.base.Mass
import pcb.uwu.amount.base.Time
import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.LengthUnit
import pcb.uwu.unit.base.TimeUnit
import pcb.uwu.unit.derived.mechanics.AccelerationUnit
import pcb.uwu.unit.derived.mechanics.ForceUnit
import pcb.uwu.unit.derived.mechanics.SpeedUnit
import pcb.uwu.util.UnitAmountUtils

open class Acceleration : CompositeUnitAmount<AccelerationUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: AccelerationUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: AccelerationUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(acceleration: UnitAmount<AccelerationUnit>) =
        Acceleration(amount = this.amount + (acceleration to this.unit).amount,
                     unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(acceleration: UnitAmount<AccelerationUnit>) =
        Acceleration(amount = this.amount - (acceleration to this.unit).amount,
                     unit = this.unit)

    override fun times(scalar: Number) =
        Acceleration(amount = this.amount * scalar,
                     unit = this.unit)

    override fun div(scalar: Number) =
        Acceleration(amount = this.amount / scalar,
                     unit = this.unit)

    override fun to(unit: AccelerationUnit) =
        Acceleration(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
                     unit = unit)

    // endregion

    // region composition

    open operator fun times(mass: Mass) =
        Force(amount = this.amount * mass.amount,
              unit = ForceUnit(accelerationUnit = this.unit,
                               massUnit = mass.unit))

    open operator fun times(time: Time) =
        Speed(amount = super.times(time).amount,
              unit = SpeedUnit(this.unit.unitCounter.findUnit(LengthUnit::class)!!,
                               this.unit.unitCounter.findUnit(TimeUnit::class)!!))

    open operator fun times(speed: Speed) =
        Time(amount = super.times(speed).amount,
             unit = this.unit.unitCounter.findUnit(TimeUnit::class)!!)

    // endregion
}
